src/txrequest.h

```diff
@@ -0,0 +1,206 @@
+     *
+     * This should be called when a transaction is successfully added to the mempool, seen in a block, or for
+     * whatever reason we no longer care about it.
+     */
+    void ForgetTxHash(const uint256& txhash);
+
+    /** Adds a new CANDIDATE announcement.
+     *
+     * Does nothing if one already exists for that (txhash, peer) combination (whether it's CANDIDATE, REQUESTED, or
+     * COMPLETED). Note that the txid/wtxid property is ignored for determining uniqueness, so if an announcement
+     * is added for a wtxid H, while one for txid H from the same peer already exists, it will be ignored. This is
+     * harmless as the txhashes being equal implies it is a non-segwit transaction, so it doesn't matter how it is
+     * fetched. The new announcement is given the specified preferred and reqtime values, and takes its is_wtxid
+     * from the specified gtxid.
+     */
+    void ReceivedInv(NodeId peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime);
+
+    /** Find the txids to request now from peer.
+     *
+     * It does the following:
+     *  - Convert all REQUESTED announcements (for all txhashes/peers) with (expiry <= now) to COMPLETED ones.
+     *  - Requestable announcements are selected: CANDIDATE announcements from the specified peer with
+     *    (reqtime <= now) for which no existing REQUESTED announcement with the same txhash exists, and for which
+     *    the specified peer is the best choice among all (reqtime <= now) CANDIDATE announcements with the same
+     *    txhash (subject to preferredness rules, and tiebreaking using a deterministic salted hash of peer and
+     *    txhash).
+     *  - The selected announcements are converted to GenTxids using their is_wtxid flag, and returned in
+     *    announcement order (even if multiple were added at the same time, or when the clock went backwards while
+     *    they were being added).
+     */
+    std::vector<GenTxid> GetRequestable(NodeId peer, std::chrono::microseconds now);
+
+    /** Marks a transaction as requested, with a specified expiry.
+     *
+     * If no CANDIDATE announcement for the provided peer and txhash exists, this call has no effect. Otherwise:
+     *  - That announcement is converted to REQUESTED.
+     *  - If any other REQUESTED announcement for the same txhash already existed, it means an unexpected request
+     *    was made (GetRequestable will never advise doing so). In this case it is converted to COMPLETED, as we're
+     *    no longer waiting for a response to it.
+     */
+    void RequestedTx(NodeId peer, const uint256& txhash, std::chrono::microseconds expiry);
+
+    /** Converts any CANDIDATE or REQUESTED announcement to a COMPLETED one, if one exists.
+     *
+     * It should be called whenever a transaction or NOTFOUND was received from a peer. When a good transaction is
+     * received, ForgetTxHash should be called instead of (or in addition to) this operation.
```

Comment:
You're right that the description doesn't match the usage in net_processing here, but "has been evaluation by the mempool" is perhaps incorrect as well. If it's seen in a block, it can be forgotten as well, for example.

Really the right criterion is: has it been made AlreadyHaveTx()==true. If we'd delete something that isn't AlreadyHaveTx(), it means we risk fetching it again if it's announced by another peer, despite already having seen it presumably. If we don't delete something that is AlreadyHaveTx(), we're risking fetching something from the same peer despite the same.

I've adapted the comments a bit and tried to formulate it in a way that doesn't depend on net_processing specifics.